server/src/main/java/org/elasticsearch/search/lookup/SearchLookup.java

```diff
@@ -27,31 +27,22 @@
 import java.util.function.Function;
 
 public class SearchLookup {
-
-    final DocLookup docMap;
-
-    final SourceLookup sourceLookup;
-
-    final FieldsLookup fieldsLookup;
+    private final DocLookup docLookup;
+    private final MapperService mapperService;
 
     public SearchLookup(MapperService mapperService, Function<MappedFieldType, IndexFieldData<?>> fieldDataLookup) {
-        docMap = new DocLookup(mapperService, fieldDataLookup);
-        sourceLookup = new SourceLookup();
-        fieldsLookup = new FieldsLookup(mapperService);
+        docLookup = new DocLookup(mapperService, fieldDataLookup);
+        this.mapperService = mapperService;
     }
 
     public LeafSearchLookup getLeafSearchLookup(LeafReaderContext context) {
         return new LeafSearchLookup(context,
-                docMap.getLeafDocLookup(context),
-                sourceLookup,
-                fieldsLookup.getLeafFieldsLookup(context));
+                docLookup.getLeafDocLookup(context),
+                new SourceLookup(),
```

Comment:
Similar to @nik9000's comment, it looks like source will be loaded twice here for cases when a script needs to access ` _source` and when `_source` is requested for a document, for example for queries like this:

```js
{
  "_source": true,
  "query": {
    "match_all": {}
  },
  "script_fields": {
    "my_script": {
      "script": {
        "source": "params['_source']['my_field']"
      }
    }
  }
}
```

Reply:
I'm not sure this is the case -- looking at `ScriptFieldsPhase`, it overwrites `hitsExecute` and so processes all hits at once. That means it wouldn't be able to re-use the loaded _source (even before the refactor) ?

Reply:
Another note: I'll add some Javadoc to `HitContext` explaining the _source sharing strategy, since it is quite tricky to understand.

Reply:
I do think this'll break the shared loading of `_source` in scripts at query time though. Each script calls this method to get its own fresh LeafSearchLookup and we sort of rely on them being able to share the loaded `_source`.

Reply:
> I do think this'll break the shared loading of _source in scripts at query time though.

👍  I agree, this will be fixed when I restore the shared `SourceLookup` on `SearchLookup`.